README.md
![20-20-eyesights](./20-20-eyesights.png)

# 20/20/20 rule

It will be opened every 20 minutes to remind you to take a break

## Features

- Set a timer for working time
- View time remaining in status bar
- Click on status bar to cancel or reset timer

## Available Commands

- 20-20-20: Start
- 20-20-20: Cancel
- 20-20-20: Restar

## Supported Settings

| Name                       | Default | Description                                                                                                                 |
| -------------------------- | ------- | --------------------------------------------------------------------------------------------------------------------------- |
| `20-20-20.workingTime`     | `true`  | Continue working should take a break for preventing eye strain. e.g. 20m for 20 minutes. 1h for 1 hour. 30s for 30 seconds. |
| `20-20-20.breakingTime`    | `true`  | Time for taking a break. e.g. 20m for 20 minutes. 1h for 1 hour. 30s for 30 seconds.                                        |
| `20-20-20.showInformation` | `true`  | Whether to show information when finish a workingTime                                                                       |
| `20-20-20.message`         | `true`  | The message to show. when finish a workingTime                                                                              |
| `20-20-20.priority`        | `true`  | The priority of the item. Higher values mean the item should be shown more to the left.                                     |

## Usage

run `Eye Strain: Stroll to grab a cup of coffee` to open Webview Panel

## 实现思路

1. 在 vscode 生命周期 onStartupFinished 执行插件
2. 插件在 StatusBar 生命周期执行，添加一个倒计时，显示格式 HH:mm
3. 创建 onCommand 命令 可以手动启动/关闭倒计时
4. 创建 Configuration 配置可以自定义:计数器的显示格式、工作倒计时时间、休息倒计时时间

## subscriptions: {dispose}[]

确保插件卸载的时候回收资源

参考 https://stackoverflow.com/questions/55554018/purpose-for-subscribing-a-command-in-vscode-extension

## About 20/20/20 rule

A helpful trick called the 20-20-20 rule for preventing eye strain. Running `Eye Strain: Stroll to grab a cup of coffee` to have a break

After or during a long day of working at a computer, have you experienced any of the following problems?

- sore, tired or burning eyes
- blurred or double vision
- watery, itchy or dry eyes
- headaches

If you have, it’s likely the result of eye strain, which happens when your eyes get tired from intense use. Fortunately, it can be remedied with a helpful trick called the 20-20-20 rule:

Every 20 minutes, look at something 20 feet away for 20 seconds.

## python script

### For Mac

Python v3.7 `test.py`

```python
import os
from threading import Timer

TIME = 60.0 * 20 # 20 min

def notify():
    title = '20/20/20 rule'
    text = 'hava a break'
    os.system("""
              osascript -e 'display notification "{}" with title "{}"'
              """.format(text, title))
    Timer(TIME, notify).start()

Timer(TIME, notify).start()
```

`nohup test.py &` run it in the background

`ps ax | grep test.py` find PID and to kill it

## Feedback

claviering@gmail.com

## Repo

[Github](https://github.com/claviering/20-20-20)

## License

[MIT License.](LICENSE.txt)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.0.0] - 2021-12-20

- show timer in status bar
- click on status bar to cancel or reset timer
- change commands

## [0.1.2] - 2020-07-13

### Add

- 用户可以自定义时间,添加计时器

### Removed

- 删除按钮

## [0.0.11] - 2019-05-02

### Add

- README

## [0.0.1] - 2019-05-02

### Add

- first version publish

### Changed

- nothing

### Removed

- nothing


src/index.js
// 前面补零
const fillZero = (num, len) => num.toString().padStart(len, "0");

/**
 * 
 * @param {Object} config 倒计时 配置
 */
export function Time(config) {
  let hour, minute, second, secondTenDigits;
  /**
   * 初始化开始的秒数
   * @param {Number | String} second 
   */
  function initSecond(second) {
    second = Number.parseInt(second) % 10;
    let secondDom = document.querySelector('#second');
    let innerHTML = '';
    for (let index = 0; index < 10; index++) {
      let num = (second - index + 10) % 10;
      innerHTML += `<div>${num}</div>`;
    }
    secondDom.innerHTML = innerHTML;
  }
  /**
   * 初始化开始的秒的十位数
   * @param {Number | String} second 
   */
  function initSecondTenDigits(secondTenDigits) {
    secondTenDigits = Number.parseInt(secondTenDigits)
    let secondTenDigitsDom = document.querySelector('.second-ten-digits');
    secondTenDigitsDom.textContent = secondTenDigits;
  }
  /**
   * 初始化开始的分钟数
   * @param {Number | String} minute 
   */
  function initMinute(minute) {
    let minuteDom = document.querySelector('.minute');
    minuteDom.textContent = fillZero(minute || '00', 2);
  }
  /**
   * 初始化开始的分钟数
   * @param {Number | String} hour 
   */
  function initHour(hour) {
    let secondTenDigits = document.querySelector('.hour');
    secondTenDigits.textContent = fillZero(hour || '00', 2);
  }
  /**
   * 初始化内容
   * @param {Number} second 
   * @param {Number} minute 
   * @param {Number} hour 
   */
  function initTextContent(second, tmpSecondTenDigits, tmpMinute, tmpHour) {
    if (second) {
      initSecond(second);
    }
    initSecondTenDigits(tmpSecondTenDigits);
    initMinute(tmpMinute);
    initHour(tmpHour);
  }
  /**
   * 秒数的十进制位减 1
   */
  function decrease() {
    secondTenDigits = secondTenDigits - 1;
    if (secondTenDigits < 0 && 0 < minute) {
      secondTenDigits = 5;
      minute -= 1;
    } else if (secondTenDigits < 0 && minute <= 0 && 0 < hour) {
      secondTenDigits = 5;
      minute = 59;
      hour -= 1;
    } else if (minute <= 0 && hour <= 0) {
      minute = 0;
    }
    initTextContent(false, secondTenDigits, minute, hour);
  }
  /**
   * 设置定时器
   * @param {Number} 总的秒数 
   */
  function interval(totalSecond) {
    if (!totalSecond) return;
    // 00 秒，立即执行一次
    if (totalSecond % 10 === 0) {
      decrease();
    }
    let timeInterval = '';
    timeInterval = setInterval(() => {
      totalSecond -= 1;
      if (!totalSecond) {
        clearInterval(timeInterval);
        togglePlay();
        return;
      }
      if (totalSecond % 10) return;
      decrease();
    }, 1000);
    togglePlay();
  }
  /**
   * 暂停/开始动画
   */
  function togglePlay() {
    let secondDom = document.querySelector('#second');
    secondDom.classList.toggle('running');
  }
  function initApp(app) {
    let appDom = document.querySelector(app);
    if (!appDom) throw new Error('app id 错误');
    appDom.innerHTML = `
    <div id="hour" class="hour">00</div>
    <div class="colon">:</div>
    <div id="minute" class="minute">00</div>
    <div class="colon">:</div>
    <div id="second-ten-digits" class="second-ten-digits">0</div>
    <div id="second" class="second play">
      <div>0</div>
      <div>9</div>
      <div>8</div>
      <div>7</div>
      <div>6</div>
      <div>5</div>
      <div>4</div>
      <div>3</div>
      <div>2</div>
      <div>1</div>
    </div>`;
  }
  /**
   * 正则表达式校验时间
   * @param {String} timer 时间
   */
  function testTimer(timer) {
    let s = /^[0-9]{2}:[0-9]{2}:[0-9]{2}/g;
    return s.test(timer);
  }
  function init() {
    const {app, timer} = config;
    if (!testTimer(timer)) {
      throw new Error('设置的时间格式错误');
    };
    let timeSplit = timer.split(':');
    hour = Number.parseInt(timeSplit[0]);
    minute = Number.parseInt(timeSplit[1]);
    second = Number.parseInt(timeSplit[2]);
    secondTenDigits = Math.floor(second / 10);
    let totalSecond = second + minute * 60 + hour * 3600;
    initApp(app);
    initTextContent(second, secondTenDigits, minute, hour);
    interval(totalSecond);
  }
  init();
}


media/main.js
// This script will be run within the webview itself
// It cannot access the main VS Code APIs directly.
(function () {
	const vscode = acquireVsCodeApi();

	const oldState = vscode.getState();

	const counter = document.getElementById('lines-of-code-counter');
	console.log(oldState);
	let currentCount = (oldState && oldState.count) || 0;
	counter.textContent = currentCount;

	setInterval(() => {
			counter.textContent = currentCount++;

			// Update state
			vscode.setState({ count: currentCount });

			// Alert the extension when the cat introduces a bug
			// if (Math.random() < Math.min(0.001 * currentCount, 0.05)) {
			// 		// Send a message back to the extension
			// 		vscode.postMessage({
			// 				command: 'alert',
			// 				text: '🐛  on line ' + currentCount
			// 		});
			// }
	}, 1000);

	// Handle messages sent from the extension to the webview
	window.addEventListener('message', event => {
			const message = event.data; // The json data that the extension sent
			switch (message.command) {
					case 'refactor':
							currentCount = Math.ceil(currentCount * 0.5);
							counter.textContent = currentCount;
							break;
			}
	});
}());

src/utils/index.ts
String.prototype.p2 = function () {
  return this.padStart(2, "0");
};
/**
 * 把时间格式化为 "ss", "mm", "mm:ss", "HH:mm", "HH:mm:ss"
 * @param format 显示时间格式
 * @param time 时间, 单位秒
 * @returns
 */
export function formateDate(format: string, time: number) {
  let { d, h, m, s, ms } = getTime(time * 1000);
  if (format === "ss") {
    return `${s + m * 60 + h * 60 * 60 + d * 24 * 60 * 60}`.p2() + "s";
  } else if (format === "mm") {
    return `${m + h * 60 + d * 24 * 60}`.p2() + "m";
  } else if (format === "mm:ss") {
    return `${m + h * 60 + d * 24 * 60}`.p2() + ":" + `${s}`.p2();
  } else if (format === "HH:mm") {
    return `${h}`.p2() + ":" + `${m + h * 60 + d * 24 * 60}`.p2();
  } else if (format === "HH:mm:ss") {
    return `${h}`.p2() + ":" + `${m}`.p2() + ":" + `${s}`.p2();
  } else {
    return format;
  }
}

// time 单位毫秒
function getTime(time: number) {
  let ms = time % 1000;
  let s = time / 1000;
  let m = s / 60;
  let h = m / 60;
  let d = h / 24;

  s = s % 60 | 0;
  m = m % 60 | 0;
  h = h % 24 | 0;
  d = d | 0;

  d = d > 0 ? d : 0;
  h = h > 0 ? h : 0;
  m = m > 0 ? m : 0;

  return { d, h, m, s, ms };
}

export function checkTimeFormat(time: string): boolean {
  let test = /^[1-9](\d*)(\.[0-9]{0,2})?[hms]$/.test(time);
  return test;
}

/**
 *
 * @param time 时间 20m, 1h, 30s
 */
export function getSecond(time: string) {
  if (!checkTimeFormat(time)) {
    return 0;
  }
  let end: string = time.slice(-1);
  let numString = time.match(/^[1-9](\d*)(\.[0-9]{0,2})?/);
  if (!numString) {
    return 0;
  }
  let num = parseFloat(numString[0]);
  if (end === "s") {
    return num;
  } else if (end === "m") {
    return num * 60;
  } else if (end === "h") {
    return num * 60 * 60;
  } else {
    return 0;
  }
}


src/extension.ts
import * as vscode from "vscode";
let myStatusBarItem: vscode.StatusBarItem;
import { getSecond, formateDate, checkTimeFormat } from "./utils";
let timer: any = null;
let time: number = 0;
let isWorking: boolean = true;
const CANCEL_TIMER_LABEL = "$(close) Cancel timer";
const RESET_TIMER_LABEL = "$(clock) Reset timer";
const ERROE_MESSAGE = 'Please input a valid time, such as "20m", "1h", "30s"';

export function activate(context: vscode.ExtensionContext) {
  const { subscriptions } = context;
  let {
    format,
    workingTime,
    breakingTime,
    showInformation,
    message,
    priority,
  } = vscode.workspace.getConfiguration("20-20-20");
  if (!checkTimeFormat(workingTime) || !checkTimeFormat(breakingTime)) {
    vscode.window.showInformationMessage(ERROE_MESSAGE);
  }
  time = getSecond(workingTime) || 20 * 60; // 默认20分钟
  let startDisposable = vscode.commands.registerCommand(
    "20-20-20.start",
    () => {
      start();
    }
  );
  let cancelDisposable = vscode.commands.registerCommand(
    "20-20-20.cancel",
    () => {
      cancel();
    }
  );
  let restartDisposable = vscode.commands.registerCommand(
    "20-20-20.restart",
    () => {
      restart();
    }
  );
  let clickDisposable = vscode.commands.registerCommand(
    "20-20-20._timer-click",
    async () => {
      const option = await vscode.window.showQuickPick(
        [CANCEL_TIMER_LABEL, RESET_TIMER_LABEL],
        { canPickMany: false }
      );
      if (!option) {
        return;
      }
      if (option === CANCEL_TIMER_LABEL) {
        cancel();
      } else if (option === RESET_TIMER_LABEL) {
        restart();
      }
    }
  );

  myStatusBarItem = vscode.window.createStatusBarItem(
    vscode.StatusBarAlignment.Right,
    priority
  );
  myStatusBarItem.command = "20-20-20._timer-click";

  function showTimer() {
    let text = formateDate(format, time);
    myStatusBarItem.text = text;
    if (time === 0 && isWorking && showInformation) {
      vscode.window.showInformationMessage(message, "Got It!");
    }
    if (time === 0) {
      time = getSecond(isWorking ? breakingTime : workingTime);
      isWorking = !isWorking;
      return;
    }
    time--;
    myStatusBarItem.show();
  }

  function start() {
    if (timer) return;
    timer = setInterval(showTimer, 1000);
    myStatusBarItem.show();
  }
  function restart() {
    isWorking = true;
    clearInterval(timer);
    time = getSecond(workingTime);
    showTimer();
    timer = setInterval(showTimer, 1000);
    myStatusBarItem.show();
  }
  function cancel() {
    if (!timer) return;
    clearInterval(timer);
    timer = null;
    myStatusBarItem.hide();
  }
  start();
  subscriptions.push(myStatusBarItem);
  subscriptions.push(startDisposable);
  subscriptions.push(cancelDisposable);
  subscriptions.push(restartDisposable);
  subscriptions.push(clickDisposable);
  subscriptions.push(
    vscode.workspace.onDidChangeConfiguration((e) => {
      if (e.affectsConfiguration("20-20-20")) {
        format = vscode.workspace.getConfiguration("20-20-20").get("format");
        workingTime = vscode.workspace
          .getConfiguration("20-20-20")
          .get("workingTime");
        breakingTime = vscode.workspace
          .getConfiguration("20-20-20")
          .get("breakingTime");
        if (!checkTimeFormat(workingTime) || !checkTimeFormat(breakingTime)) {
          vscode.window.showInformationMessage(ERROE_MESSAGE);
        } else {
          time = getSecond(isWorking ? workingTime : breakingTime);
        }
        showInformation = vscode.workspace
          .getConfiguration("20-20-20")
          .get("showInformation");
        message = vscode.workspace.getConfiguration("20-20-20").get("message");
        let newPriority: number = vscode.workspace
          .getConfiguration("20-20-20")
          .get("priority") as number;
        if (newPriority !== priority) {
          myStatusBarItem.dispose();
          myStatusBarItem = vscode.window.createStatusBarItem(
            vscode.StatusBarAlignment.Right,
            newPriority
          );
          subscriptions.push(myStatusBarItem);
        }
      }
    })
  );
}

export function deactivate() {
  clearInterval(timer);
}
